cmd/slashland/main.go

```diff
@@ -0,0 +1,354 @@
+	userID   string
+	userName string
+	ref      string
+	private  bool
+}
+
+func main() {
+	log.SetFlags(log.Lshortfile)
+	env.Parse()
+
+	http.HandleFunc("/slash", slashLand)
+	http.HandleFunc("/health", func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(200)
+	})
+
+	go lander()
+
+	say("Ready for action!")
+	err := http.ListenAndServe(":"+*port, nil)
+	log.Fatalln(err)
+}
+
+func slashLand(w http.ResponseWriter, r *http.Request) {
+	if r.FormValue("token") != *slackToken {
+		http.Error(w, "unauthorized", http.StatusUnauthorized)
+		return
+	}
+	if !contains(r.FormValue("channel_id"), *slackChannels) {
+		http.Error(w, fmt.Sprintf("%s not enabled in this channel.", r.FormValue("command")), 400)
+		return
+	}
+	a := strings.Fields(r.FormValue("text"))
+	private := false
+	if len(a) >= 1 && a[0] == "-prv" {
+		private = true
+		a = a[1:]
+	}
+	if len(a) != 1 {
+		http.Error(w, slashLandUsage, 422)
+		return
+	}
+	landReqs <- &landReq{
+		ref:      a[0],
+		userID:   r.FormValue("user_id"),
+		userName: r.FormValue("user_name"),
+		private:  private,
+	}
+	sayf("<@%s|%s> is attempting to land %s",
+		r.FormValue("user_id"),
+		r.FormValue("user_name"),
+		a[0],
+	)
+}
+
+func lander() {
+	for req := range landReqs {
+		land(req)
+	}
+}
+
+func land(req *landReq) {
+	defer catch()
+
+	repo := *repo
+	if req.private {
+		repo = *privRepo
+	}
+
+	gopath := "/tmp/land"
+	landdir := gopath + "/src/" + repo
+
+	fetch(landdir, req.ref, repo)
+	commit := string(bytes.TrimSpace(runOutput(landdir, exec.Command("git", "rev-parse", "HEAD"))))
+
+	prBits, err := pipeline(
+		dirCmd(landdir, "git", "ls-remote", "origin", `refs/pull/*/head`),
+		exec.Command("fgrep", commit),
+		exec.Command("cut", "-d/", "-f3"),
+	)
+	if err != nil {
+		sayf("<@%s|%s> failed to land %s: could not find open pull request",
+			req.userID,
+			req.userName,
+			req.ref,
+		)
+		return
+	}
+	pr := string(bytes.TrimSpace(prBits))
+
+	var prState struct {
+		Title     string
+		Body      string
+		Merged    bool
+		Mergeable *bool
+	}
+	err = doGithubReq("GET", "repos/"+*org+"/"+repo+"/pulls/"+pr, nil, &prState)
+	if err != nil {
+		sayf("<@%s|%s> failed to land %s: error fetching github status",
+			req.userID,
+			req.userName,
+			req.ref,
+		)
+	}
+	if prState.Merged {
+		sayf("<@%s|%s> %s has already landed",
+			req.userID,
+			req.userName,
+			req.ref,
+		)
+		return
+	}
+	if prState.Mergeable != nil && *prState.Mergeable == false {
+		sayf("<@%s|%s> failed to land %s: branch has conflicts",
+			req.userID,
+			req.userName,
+			req.ref,
+		)
+		return
+	}
+
+	cmd := dirCmd(landdir, "git", "rebase", "origin/main")
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	if err != nil {
+		cmd = dirCmd(landdir, "git", "rebase", "--abort")
+		cmd.Stderr = os.Stderr
+		err = cmd.Run()
+		sayf("<@%s|%s> failed to land %s: branch has conflicts",
+			req.userID,
+			req.userName,
+			req.ref,
+		)
+		return
+	}
+
+	cmd = dirCmd(landdir, "git", "filter-branch", "--env-filter", `
```

Comment:
For the actual filter, does that need to contain single quotes, as it would run on the command line?